arrow-cast/src/parse.rs

```diff
@@ -132,6 +132,106 @@ pub fn string_to_timestamp_nanos(s: &str) -> Result<i64, ArrowError> {
     )))
 }
 
+/// Accepts a string in ISO8601 standard format and some
+/// variants and converts it to nanoseconds since midnight.
+///
+/// Examples of accepted inputs:
+/// * `09:26:56.123 AM`
+/// * `23:59:59`
+/// * `6:00 pm`
+//
+/// Internally, this function uses the `chrono` library for the
+/// time parsing
+///
+/// ## Timezone / Offset Handling
+///
+/// This function does not support parsing strings with a timezone
+/// or offset specified, as it considers only time since midnight.
+pub fn string_to_time_nanoseconds(s: &str) -> Result<i64, ArrowError> {
+    // colon count, presence of decimal, presence of whitespace
+    fn preprocess_time_string(string: &str) -> (u8, bool, bool) {
+        string
+            .chars()
+            .fold((0, false, false), |tup, char| match char {
+                ':' => (tup.0.saturating_add(1), tup.1, tup.2),
```

Comment:
```suggestion
            .fold((0_usize, false, false), |tup, char| match char {
                ':' => (tup.0 + 1, tup.1, tup.2),
```

Using a usize means this can't actually overflow